File: core/src/main/kotlin/com/example/toyracers/race/RaceRules.kt
package com.example.toyracers.race

import com.example.toyracers.track.StartLine
import com.example.toyracers.track.Track
import com.example.toyracers.track.TrackPoint
import com.example.toyracers.track.TrackSegment
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Advances ordered checkpoint, lap, and timing state from a car's simulated movement.
 *
 * Callers can suppress progression for teleports and respawns while still advancing race time.
 */
class RaceRules(
    private val track: Track,
    val requiredLaps: Int = DEFAULT_LAP_COUNT,
) {
    private var nextFinishPosition = 1

    init {
        require(requiredLaps > 0) { "Required lap count must be positive" }
    }

    fun update(
        progress: RaceProgress,
        previousPosition: TrackPoint,
        currentPosition: TrackPoint,
        deltaSeconds: Float,
        allowProgress: Boolean = true,
    ) {
        require(deltaSeconds >= 0f) { "Delta time must not be negative" }
        if (progress.finished) return

        progress.totalRaceTime += deltaSeconds
        if (!allowProgress) return

        if (progress.currentCheckpointIndex < track.checkpoints.size) {
            val checkpoint = track.checkpoints[progress.currentCheckpointIndex]
            if (
                crossesForward(
                    previousPosition,
                    currentPosition,
                    checkpoint.gate,
                    checkpoint.forwardX,
                    checkpoint.forwardY,
                )
            ) {
                progress.currentCheckpointIndex++
            }
            return
        }

        if (!crossesForward(previousPosition, currentPosition, track.startLine)) return

        progress.completedLaps++
        val lapTime = progress.totalRaceTime - progress.lapStartTime
        progress.bestLapTime = progress.bestLapTime?.let { minOf(it, lapTime) } ?: lapTime
        progress.lapStartTime = progress.totalRaceTime
        progress.currentCheckpointIndex = 0

        if (progress.completedLaps >= requiredLaps) {
            progress.finished = true
            progress.finishPosition = nextFinishPosition++
        }
    }

    private fun crossesForward(
        previous: TrackPoint,
        current: TrackPoint,
        startLine: StartLine,
    ): Boolean {
        val centerX = startLine.bounds.x + startLine.bounds.width / 2f
        val centerY = startLine.bounds.y + startLine.bounds.height / 2f
        val forwardLength = vectorLength(startLine.forwardX, startLine.forwardY)
        val perpendicularX = -startLine.forwardY / forwardLength
        val perpendicularY = startLine.forwardX / forwardLength
        val halfLength =
            abs(perpendicularX) * startLine.bounds.width / 2f +
                abs(perpendicularY) * startLine.bounds.height / 2f
        val gate = TrackSegment(
            start = TrackPoint(
                centerX - perpendicularX * halfLength,
                centerY - perpendicularY * halfLength,
            ),
            end = TrackPoint(
                centerX + perpendicularX * halfLength,
                centerY + perpendicularY * halfLength,
            ),
        )
        return crossesForward(
            previous,
            current,
            gate,
            startLine.forwardX,
            startLine.forwardY,
        )
    }

    private fun crossesForward(
        previous: TrackPoint,
        current: TrackPoint,
        gate: TrackSegment,
        forwardX: Float,
        forwardY: Float,
    ): Boolean {
        val forwardLength = vectorLength(forwardX, forwardY)
        val normalX = forwardX / forwardLength
        val normalY = forwardY / forwardLength
        val previousDistance =
            (previous.x - gate.start.x) * normalX +
                (previous.y - gate.start.y) * normalY
        val currentDistance =
            (current.x - gate.start.x) * normalX +
                (current.y - gate.start.y) * normalY
        if (previousDistance >= -CROSSING_EPSILON || currentDistance < -CROSSING_EPSILON) {
            return false
        }

        val movementX = current.x - previous.x
        val movementY = current.y - previous.y
        if (movementX * normalX + movementY * normalY <= CROSSING_EPSILON) return false

        val crossingFraction = previousDistance / (previousDistance - currentDistance)
        val crossingX = previous.x + movementX * crossingFraction
        val crossingY = previous.y + movementY * crossingFraction
        val gateX = gate.end.x - gate.start.x
        val gateY = gate.end.y - gate.start.y
        val gateLengthSquared = gateX * gateX + gateY * gateY
        val gateFraction =
            ((crossingX - gate.start.x) * gateX + (crossingY - gate.start.y) * gateY) /
                gateLengthSquared
        return gateFraction in -CROSSING_EPSILON..(1f + CROSSING_EPSILON)
    }

    private fun vectorLength(
        x: Float,
        y: Float,
    ): Float = sqrt(x * x + y * y)

    companion object {
        const val DEFAULT_LAP_COUNT = 3
        private const val CROSSING_EPSILON = 0.0001f
    }
}

File: dart/lib/simulation/race/race_rules.dart
import 'dart:math' as math;

import '../math/float32.dart';
import '../track/track.dart';
import '../track/track_geometry.dart';
import '../track/track_point.dart';
import 'race_progress.dart';

/// Advances ordered checkpoint, lap, and timing state from simulated movement.
///
/// Callers can disable progression for a teleport or respawn while retaining
/// timer advancement. This intentionally preserves the Kotlin rule behaviour,
/// including its single-gate-per-update progression.
final class RaceRules {
  RaceRules(this._track, {this.requiredLaps = defaultLapCount}) {
    if (requiredLaps <= 0) {
      throw ArgumentError.value(
        requiredLaps,
        'requiredLaps',
        'must be positive',
      );
    }
  }

  final Track _track;
  final int requiredLaps;
  int _nextFinishPosition = 1;

  /// Synchronizes a restored fixture's finish state before further updates.
  void synchronizeFinishOrdering(Iterable<RaceProgress> progresses) {
    final finishPositions = <int>[];
    for (final progress in progresses) {
      if (progress.finished) {
        final finishPosition = progress.finishPosition;
        if (finishPosition == null) {
          throw ArgumentError('Finished progress must have a finish position');
        }
        finishPositions.add(finishPosition);
      }
    }
    if (finishPositions.toSet().length != finishPositions.length) {
      throw ArgumentError('Finish positions must be unique');
    }
    if (finishPositions.isEmpty) {
      return;
    }
    final highestPosition = finishPositions.reduce(math.max);
    if (highestPosition == _maximumInt) {
      throw ArgumentError(
        'Finish position is too large to assign another result',
      );
    }
    _nextFinishPosition = math.max(_nextFinishPosition, highestPosition + 1);
  }

  /// Clears finish ordering before a new race starts.
  void resetFinishOrdering() => _nextFinishPosition = 1;

  /// Applies one participant's elapsed time and possible gate crossing.
  void update({
    required RaceProgress progress,
    required TrackPoint previousPosition,
    required TrackPoint currentPosition,
    required double deltaSeconds,
    bool allowProgress = true,
  }) {
    final delta = Float32.narrow(deltaSeconds);
    if (delta < 0) {
      throw ArgumentError.value(
        deltaSeconds,
        'deltaSeconds',
        'must not be negative',
      );
    }
    if (progress.finished) {
      return;
    }

    progress.totalRaceTime = Float32.add(progress.totalRaceTime, delta);
    if (!allowProgress) {
      return;
    }

    if (progress.currentCheckpointIndex < _track.checkpoints.length) {
      final checkpoint = _track.checkpoints[progress.currentCheckpointIndex];
      if (_crossesForwardGate(
        previous: previousPosition,
        current: currentPosition,
        gate: checkpoint.gate,
        forwardX: checkpoint.forwardX,
        forwardY: checkpoint.forwardY,
      )) {
        progress.currentCheckpointIndex++;
      }
      return;
    }

    if (!_crossesForwardStartLine(previousPosition, currentPosition)) {
      return;
    }
    _completeLap(progress);
  }

  void _completeLap(RaceProgress progress) {
    progress.completedLaps++;
    final lapTime = Float32.subtract(
      progress.totalRaceTime,
      progress.lapStartTime,
    );
    final bestLapTime = progress.bestLapTime;
    progress.bestLapTime = bestLapTime == null || lapTime < bestLapTime
        ? lapTime
        : bestLapTime;
    progress.lapStartTime = progress.totalRaceTime;
    progress.currentCheckpointIndex = 0;

    if (progress.completedLaps >= requiredLaps) {
      progress.finished = true;
      progress.finishPosition = _nextFinishPosition++;
    }
  }

  bool _crossesForwardStartLine(TrackPoint previous, TrackPoint current) {
    final startLine = _track.startLine;
    final centerX = Float32.add(
      startLine.bounds.x,
      Float32.divide(startLine.bounds.width, 2),
    );
    final centerY = Float32.add(
      startLine.bounds.y,
      Float32.divide(startLine.bounds.height, 2),
    );
    final forwardLength = _vectorLength(startLine.forwardX, startLine.forwardY);
    final perpendicularX = Float32.divide(-startLine.forwardY, forwardLength);
    final perpendicularY = Float32.divide(startLine.forwardX, forwardLength);
    final halfLength = Float32.add(
      Float32.divide(
        Float32.multiply(perpendicularX.abs(), startLine.bounds.width),
        2,
      ),
      Float32.divide(
        Float32.multiply(perpendicularY.abs(), startLine.bounds.height),
        2,
      ),
    );
    final gate = TrackSegment(
      TrackPoint(
        Float32.subtract(centerX, Float32.multiply(perpendicularX, halfLength)),
        Float32.subtract(centerY, Float32.multiply(perpendicularY, halfLength)),
      ),
      TrackPoint(
        Float32.add(centerX, Float32.multiply(perpendicularX, halfLength)),
        Float32.add(centerY, Float32.multiply(perpendicularY, halfLength)),
      ),
    );
    return _crossesForwardGate(
      previous: previous,
      current: current,
      gate: gate,
      forwardX: startLine.forwardX,
      forwardY: startLine.forwardY,
    );
  }

  bool _crossesForwardGate({
    required TrackPoint previous,
    required TrackPoint current,
    required TrackSegment gate,
    required double forwardX,
    required double forwardY,
  }) {
    final forwardLength = _vectorLength(forwardX, forwardY);
    final normalX = Float32.divide(forwardX, forwardLength);
    final normalY = Float32.divide(forwardY, forwardLength);
    final previousDistance = _dotFromGate(previous, gate, normalX, normalY);
    final currentDistance = _dotFromGate(current, gate, normalX, normalY);
    if (previousDistance >= -_crossingEpsilon ||
        currentDistance < -_crossingEpsilon) {
      return false;
    }

    final movementX = Float32.subtract(current.x, previous.x);
    final movementY = Float32.subtract(current.y, previous.y);
    final movementAlongNormal = Float32.add(
      Float32.multiply(movementX, normalX),
      Float32.multiply(movementY, normalY),
    );
    if (movementAlongNormal <= _crossingEpsilon) {
      return false;
    }

    final crossingFraction = Float32.divide(
      previousDistance,
      Float32.subtract(previousDistance, currentDistance),
    );
    final crossingX = Float32.add(
      previous.x,
      Float32.multiply(movementX, crossingFraction),
    );
    final crossingY = Float32.add(
      previous.y,
      Float32.multiply(movementY, crossingFraction),
    );
    final gateX = Float32.subtract(gate.end.x, gate.start.x);
    final gateY = Float32.subtract(gate.end.y, gate.start.y);
    final gateLengthSquared = Float32.add(
      Float32.multiply(gateX, gateX),
      Float32.multiply(gateY, gateY),
    );
    final gateFraction = Float32.divide(
      Float32.add(
        Float32.multiply(Float32.subtract(crossingX, gate.start.x), gateX),
        Float32.multiply(Float32.subtract(crossingY, gate.start.y), gateY),
      ),
      gateLengthSquared,
    );
    return gateFraction >= -_crossingEpsilon &&
        gateFraction <= Float32.add(1, _crossingEpsilon);
  }

  double _dotFromGate(
    TrackPoint point,
    TrackSegment gate,
    double normalX,
    double normalY,
  ) => Float32.add(
    Float32.multiply(Float32.subtract(point.x, gate.start.x), normalX),
    Float32.multiply(Float32.subtract(point.y, gate.start.y), normalY),
  );

  double _vectorLength(double x, double y) => Float32.narrow(
    math.sqrt(Float32.add(Float32.multiply(x, x), Float32.multiply(y, y))),
  );

  static const int defaultLapCount = 3;
  static const int _maximumInt = 2147483647;
  static final double _crossingEpsilon = Float32.narrow(0.0001);
}
